contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,94 @@
+# Apply v2
+
+## Background
+
+`kubectl apply` reads a file or set of files, and updates the cluster state based off the file contents.
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+Reducing the code complexity and improving the correctness will also make it simpler to move this logic
+to the server as part of an API in the future, while also addressing the immediate needs of the client.
+
+## Changes
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  `--dry-run` is nearly useless for understanding the
```

Comment:
Why is dry run useless? Because it is also buggy and incomplete? Because it is implemented in the client rather than in the server?

Reply:
It doesn't show the user what they really want to know - what will be the new object in the cluster after apply is run.  It somewhat shows the changes being made, but without the context of what they are being made on.

- I would expect apply users care less about the patch format itself, and care more about the object in the cluster that will result from the patch.
- The patch output obfuscates validation errors caused by the values set by patch conflicting with other values set on the live object missing from the config (e.g. rolloutStrategy for Deployment)
- The patch itself is buggy and broken cases are hard to see just by looking at a patch (e.g. container ports merge key being incorrect)
- The server may apply the patch differently than the client, so running the patch locally against the file may result in something different than what the server does (e.g. the server doesn't support optional directives because it is skewed)

Reply:
FWIW, I don't understand what apply's current patch shows me when I do `--dry-run -o json`.  It looks like it is just showing me the local object without applying the patch.